packages/sync/src/cache/createMutationOptions.ts

```diff
@@ -0,0 +1,125 @@
+import { MutationOptions, OperationVariables, MutationUpdaterFn } from "apollo-client";
+import { DocumentNode } from "graphql";
+import { CacheOperation } from "./CacheOperation";
+import { createOptimisticResponse } from "./createOptimisticResponse";
+import { Query, QueryWithVariables } from "./CacheUpdates";
+import { getOperationFieldName } from "../utils/helpers";
+
+export interface MutationHelperOptions {
+  mutation: DocumentNode;
+  variables: OperationVariables;
+  updateQuery: Query;
+  typeName: string;
+  operationType?: CacheOperation;
+  idField?: string;
+}
+
+export const createMutationOptions = (options: MutationHelperOptions): MutationOptions => {
+  const {
+    mutation,
+    variables,
+    updateQuery,
+    typeName,
+    operationType = CacheOperation.ADD,
+    idField = "id"
+  } = options;
+  const operationName = getOperationFieldName(mutation);
+  const optimisticResponse = createOptimisticResponse(operationName,
+    typeName,
+    variables,
+    operationType === CacheOperation.ADD);
+
+  const update = getUpdateFunction(operationName, idField, updateQuery, operationType);
+  return { mutation, variables, optimisticResponse, update };
+};
+
+// returns the update function used to update the cache by the client
+// ignores the scenario where the cache operation is an update as this is handled automatically
+// from Apollo client 2.5 onwards.
+const getUpdateFunction = (
+  operation: string,
+  idField: string,
+  updateQuery: Query,
+  opType: CacheOperation): MutationUpdaterFn => {
+
+  const { query, variables } = deconstructQuery(updateQuery);
+  const queryField = getOperationFieldName(query);
+
+  let updateFunction: MutationUpdaterFn;
+
+  switch (opType) {
+    case CacheOperation.ADD:
+      updateFunction = (cache, { data }) => {
+        try {
+          if (data) {
+            let queryResult = cache.readQuery({ query, variables }) as any;
+            const operationData = data[operation];
+            const result = queryResult[queryField];
+            if (result) {
+              if (!result.find((item: any) => {
+                return item[idField] === operationData[idField];
+              })) {
+                result.push(operationData);
+              }
+            } else {
+              queryResult = [result];
+            }
+            cache.writeQuery({
+              query,
+              variables,
+              data: queryResult
+            });
+          }
+        } catch (e) {
+          throw (e);
```

Comment:
Caching and throwing the same error is pointless. Originally I left that empty
because it was crashing for me on Android on application startup for the first time when cache was empty)